src/lara-app/components/runtime.tsx

```diff
@@ -79,34 +84,50 @@ export const RuntimeComponent = ({experiment, runKey, firebaseJWT, setError, def
 
   // get experiment data when loaded
   useEffect(() => {
-    firebase
-      .firestore()
-      .collection(`runs/${runKey}/experiments`)
-      .orderBy("createdAt", "desc")
-      .limit(1)
-      .onSnapshot(snapshot => {
-        setQueriedFirestore(true);
-        experimentRef.current = snapshot.docs[0]?.data?.().experiment;
-        const newData = experimentRef.current?.data as IExperimentData | undefined;
-        setExperimentData(newData);
-
-        // if there is no data force an upload
-        setDisplayQrAndMaybeRegenerateQR((newData === undefined) && !reportMode);
-      }, (err) => {
-        setError(err);
-      });
+    if (runKey) {
+      firebase
+        .firestore()
+        .collection(`runs/${runKey}/experiments`)
+        .orderBy("createdAt", "desc")
+        .limit(1)
+        .onSnapshot(snapshot => {
+          setQueriedFirestore(true);
+          experimentRef.current = snapshot.docs[0]?.data?.().experiment;
+          const newData = experimentRef.current?.data as IExperimentData | undefined;
+          setExperimentData(newData);
+
+          // if there is no data force an upload
+          setDisplayQrAndMaybeRegenerateQR((newData === undefined) && !reportOrPreviewMode);
+        }, (err) => {
+          setError(err);
+        });
+    }
   }, []);
 
   // re-generate QR code if showing
   useEffect(() => {
     const interval = setInterval(() => {
-      if (displayingCode.current && !reportMode) {
-        generateQRCode();
+      if (displayingCode.current && !reportOrPreviewMode && runKey && firebaseJWT) {
+        generateQRCode({runKey, firebaseJWT});
       }
     }, UPDATE_QR_INTERVAL);
     return () => clearInterval(interval);
   }, []);
 
+  useEffect(() => {
+    if (containerRef.current) {
+      const checkHeight = () => {
+        const {height} = containerRef.current ? containerRef.current.getBoundingClientRect() : { height: 0 };
+        if (height > 0) {
+          setHeight(height);
+        }
+      };
+
+      const ro = new ResizeObserver(checkHeight);
+      ro.observe(containerRef.current);
+    }
+  }, [containerRef.current]);
```

Comment:
Cleanup function with `ro.disconnect()` could be useful (especially if it gets refactored to a hook as Kirk suggests).

Reply:
There's another related (possibly theoretical) concern in that the stated dependency on `containerRef.current` implies that the container can be changed, in which case presumably any previous observer should be disconnected before a new observer is connected. I don't know whether that's a scenario we need to handle in practice, however. It came up when Chris and I were working with the `useAutoHeight()` hook in the question-interactives repository, which has the same issue.

Reply:
React will call the previous cleanup function before re-executing this callback. ResizeObserver doesn't take any arguments in `disconnect()` function. So, I think the cleanup function should handle this case even if `containerRef.current` has already a new value. 

Even more theoretical issue is that change of the ref won't necessarily trigger react re-render. But since it's a DOM ref, I think it actually means that things just got re-rendered. So, I think it's fine in practice.

Reply:
I've added a disconnect call on the returned function from the useEffect.